### PR TITLE
Issue/2000 update new renderables

### PR DESCRIPTION
**Pull request by ylvaselling**
To  be used with this PR https://github.com/OpenSpace/OpenSpace/pull/3289
See that PR for instructions

Question to be discussed: how do we want to structure our Middleware vs Reducers? Now for example we have a dispatch to refresh groups in two different places, but I didn't really know how to organize it (what is the custom here?)

### CODE REVIEW

File: src/api/Middleware/events.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      break;
+    }
+    case "PropertyOwnerRemoved": {
+      store.dispatch(removePropertyOwners({ uris: [ data.Uri ] }));
+      store.dispatch(refreshGroups());
+      break;
+    }
+    default: {
+
+    }
```
Comment (alexanderbock) [inline src/api/Middleware/events.js:21]:
```suggestion
    case "PropertyOwnerAdded":
      store.dispatch(addPropertyOwner({ uri: data.Uri }));
      break;
    case "PropertyOwnerRemoved":
      store.dispatch(removePropertyOwners({ uris: [ data.Uri ] }));
      store.dispatch(refreshGroups());
      break;
    default:
      break;
```

File: src/api/Middleware/events.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    // Start subscribing to all events  
+    eventTopic = api.startTopic('event', {
+        event: '*',
+        status: 'start_subscription'
+    });
+
+    // eslint-disable-next-line no-restricted-syntax
+    for await (const data of eventTopic.iterator()) {
+        handleData(store, data);
+    }
```
Comment (alexanderbock) [inline src/api/Middleware/events.js:45]:
```suggestion
  // Start subscribing to all events  
  eventTopic = api.startTopic('event', {
    event: '*',
    status: 'start_subscription'
  });

  // eslint-disable-next-line no-restricted-syntax
  for await (const data of eventTopic.iterator()) {
    handleData(store, data);
  }
```

File: src/api/Middleware/events.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+const events = (store) => (next) => (action) => {
+  const result = next(action);
+  const state = store.getState();
+  switch (action.type) {
+    case actionTypes.onOpenConnection:
+        if (nSubscribers === 0) {
```
Comment (alexanderbock) [inline src/api/Middleware/events.js:53]:
```suggestion
      if (nSubscribers === 0) {
```

File: src/api/Middleware/events.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        status: 'start_subscription'
+    });
+
+    // eslint-disable-next-line no-restricted-syntax
+    for await (const data of eventTopic.iterator()) {
+        handleData(store, data);
+    }
+}
+
+const events = (store) => (next) => (action) => {
```
Comment (alexanderbock) [inline src/api/Middleware/events.js:48]: Javascript looks funky sometimes :D

File: src/api/Reducers/propertyTree.js [hunk 5]
```diff
@@ -68,13 +68,41 @@ const propertyOwners = (state = {}, action = {}) => {
           subowners: owner.subowners,
           tags: owner.tags || []
         };
+        
+        // Ensure the parents of the uri have the links to the new entry
+        // Get parent uri
+        var periodPos = owner.uri.lastIndexOf('.');
```
Comment (alexanderbock) [inline src/api/Reducers/propertyTree.js:74]:
```suggestion
        let periodPos = owner.uri.lastIndexOf('.');
```

File: src/api/Reducers/propertyTree.js [hunk 6]
```diff
@@ -68,13 +68,41 @@ const propertyOwners = (state = {}, action = {}) => {
...
+        var parentUri = owner.uri.substring(0, periodPos);
```
Comment (alexanderbock) [inline src/api/Reducers/propertyTree.js:75]:
```suggestion
        let parentUri = owner.uri.substring(0, periodPos);
```

File: src/api/Reducers/propertyTree.js [hunk 7]
```diff
@@ -68,13 +68,41 @@ const propertyOwners = (state = {}, action = {}) => {
...
+
+        // If that parent exists and the link doesn't exist, add the link to the parent
+        if (parentUri && newState[parentUri] && !newState[parentUri].subowners.includes(owner.uri)) {
+          newState[parentUri].subowners.push(owner.uri);
```
Comment (alexanderbock) [inline src/api/Reducers/propertyTree.js:79]:
```suggestion
        if (parentUri && newState[parentUri] && !newState[parentUri].subowners.includes(owner.uri))
        {
          newState[parentUri].subowners.push(owner.uri);
```

File: src/api/Reducers/propertyTree.js [hunk 8]
```diff
@@ -68,13 +68,41 @@ const propertyOwners = (state = {}, action = {}) => {
...
     case actionTypes.removePropertyOwners: {
-      const newState = { ...state };
+      let newState = { ...state };
+      
       action.payload.uris.forEach((uri) => {
+        // Delete this particular property owner
         delete newState[uri];
+
+        // Delete the parent's link to the property owner 
+        var periodPos = uri.lastIndexOf('.');
```
Comment (alexanderbock) [inline src/api/Reducers/propertyTree.js:92]:
```suggestion
        let periodPos = uri.lastIndexOf('.');
```

File: src/api/Reducers/propertyTree.js [hunk 9]
```diff
@@ -68,13 +68,41 @@ const propertyOwners = (state = {}, action = {}) => {
...
+        var parentUri = uri.substring(0, periodPos);
```
Comment (alexanderbock) [inline src/api/Reducers/propertyTree.js:93]:
```suggestion
        let parentUri = uri.substring(0, periodPos);
```

File: src/api/Reducers/propertyTree.js [hunk 10]
```diff
@@ -68,13 +68,41 @@ const propertyOwners = (state = {}, action = {}) => {
...
+        if (index > -1) { 
+          // 2nd parameter means remove one item only
+          newState[parentUri].subowners.splice(index, 1); 
+        }
+
+        // Delete subowners that have been flattened
+        const related = Object.keys(newState).filter(value => value.includes(`${uri}.`));
+        related.forEach((subOwnerUri) => 
+          delete newState[subOwnerUri]
+        );
```
Comment (alexanderbock) [inline src/api/Reducers/propertyTree.js:105]:
```suggestion
        related.forEach((subOwnerUri) => delete newState[subOwnerUri]);
```

Any specific reason for it to not be on one row?

File: src/components/BottomBar/ScreenSpaceRenderablePanel.jsx [hunk 11]
```diff
@@ -24,12 +24,17 @@ function ScreenSpaceRenderablePanel() {
...
-    (state) => (state.propertyTree.propertyOwners.ScreenSpace ?
-      state.propertyTree.propertyOwners.ScreenSpace.subowners :
-      [])
+  // Access the propertyOwners state so that useSelector is triggered by updates
+  const propertyOwners = useSelector(
+    (state) => {
+      if (state.propertyTree.propertyOwners) {
+        return state.propertyTree.propertyOwners;
+      }
+        return [];
```
Comment (alexanderbock) [inline src/components/BottomBar/ScreenSpaceRenderablePanel.jsx:33]:
```suggestion
     return state.propertyTree.propertyOwners ? state.propertyTree.propertyOwners : [];
```

File: src/api/Reducers/propertyTree.js (hunk 5 above)
Comment (WeirdRubberDuck) [reply src/api/Reducers/propertyTree.js:74]: could be `const`? (same for the comments below)

File: src/api/Reducers/propertyTree.js (hunk 8 above)
Comment (WeirdRubberDuck) [reply src/api/Reducers/propertyTree.js:92]: same here, could be `const`

File: src/components/Sidebar/Properties/PropertyOwner.jsx [hunk 12]
```diff
@@ -24,6 +23,111 @@ import ToggleContent from '../../common/ToggleContent/ToggleContent';
 import Property from './Property';
 import PropertyOwnerHeader from './PropertyOwnerHeader';
 
+// Render draggable and reorderable list with layers, using Beautinful DnD
+// Based on video tutorial: https://www.youtube.com/watch?v=aYZRRyukuIw&ab_channel=ColbyFayock
+function DragDropLayerList({ expansionIdentifier, uri }) {
+  const [trigger, setTrigger] = React.useState(true);
```
Comment (WeirdRubberDuck) [inline src/components/Sidebar/Properties/PropertyOwner.jsx:29]: The `trigger` does not seem to be used

File: src/api/Middleware/propertyTree.js [hunk 13]
```diff
@@ -126,50 +126,53 @@ const markAllSubscriptionsAsPending = () => {
...
+const addPropertyOwner = async (dispatch, uri) => {
+  const value = await api.getProperty(uri);
+  if (!value) {
+    console.error("Error retrieving property with uri ", uri);
+    return;
+  }
 
+  // Extract the data from the property owner
   const { propertyOwners, properties } = flattenPropertyTree(value);
+
```
Comment (WeirdRubberDuck) [inline src/api/Middleware/propertyTree.js:175]: Here is some more trailing whitespace

File: src/api/Middleware/propertyTree.js [hunk 14]
```diff
@@ -126,50 +126,53 @@ const markAllSubscriptionsAsPending = () => {
...
+    properties
   };
 };
 
-const getPropertyTree = async (dispatch) => {
-  const value = await api.getProperty(rootOwnerKey);
+const addPropertyOwner = async (dispatch, uri) => {
+  const value = await api.getProperty(uri);
+  if (!value) {
+    console.error("Error retrieving property with uri ", uri);
```
Comment (WeirdRubberDuck) [inline src/api/Middleware/propertyTree.js:169]:
```suggestion
    console.error('Error retrieving property with uri ', uri);
```
According to our linting settings

File: src/api/Reducers/propertyTree.js [hunk 15]
```diff
@@ -68,13 +68,41 @@ const propertyOwners = (state = {}, action = {}) => {
...
+        var parentUri = uri.substring(0, periodPos);
+        const index = newState[parentUri].subowners.indexOf(uri);
+        // If found, remove
+        if (index > -1) { 
+          // 2nd parameter means remove one item only
+          newState[parentUri].subowners.splice(index, 1); 
+        }
+
+        // Delete subowners that have been flattened
+        const related = Object.keys(newState).filter(value => value.includes(`${uri}.`));
```
Comment (WeirdRubberDuck) [inline src/api/Reducers/propertyTree.js:102]:
```suggestion
        const related = Object.keys(newState).filter((value) => value.includes(`${uri}.`));
```

File: src/api/Reducers/propertyTree.js [hunk 16]
```diff
@@ -59,7 +59,7 @@ const propertyOwners = (state = {}, action = {}) => {
   switch (action.type) {
     case actionTypes.addPropertyOwners: {
       const inputOwners = action.payload.propertyOwners;
-      const newState = { ...state };
+      let newState = { ...state };
```
Comment (WeirdRubberDuck) [inline src/api/Reducers/propertyTree.js:62]:
```suggestion
      const newState = { ...state };
```
According to linting. Even if we add things to the object further down, it does not change the reference of the object and hence it should still be const, IIRC

File: src/api/Middleware/propertyTree.js [hunk 17]
```diff
@@ -195,8 +198,12 @@ const propertyTree = (store) => (next) => (action) => {
       store.dispatch(reloadPropertyTree());
       break;
     }
+    case actionTypes.addPropertyOwner: {
+      addPropertyOwner(store.dispatch, action.payload.uri)
```
Comment (WeirdRubberDuck) [inline src/api/Middleware/propertyTree.js:202]:
```suggestion
      addPropertyOwner(store.dispatch, action.payload.uri);
```

**Review by WeirdRubberDuck — CHANGES_REQUESTED** [review]
There are quite a few eslint errors. Please enable and fix all problems in the file, as well as remove trailing whitespaces

I don't really have a good answer for the middleware VS reducer structures. My impression is that the middleware could really do anything, and the reducer should update the state. But it was a very long time since I looked at any of this 😅 and as you say, I doubt that we are even consistent in the first place. 

Worth to look at in the UI rewrite, maybe?

File: src/components/Sidebar/Properties/PropertyOwner.jsx (hunk 12 above)
Comment (ylvaselling) [reply src/components/Sidebar/Properties/PropertyOwner.jsx:29]: No, the trigger is when the "trigger" is set with the setTrigger function... it makes the component re-render

**Pull request merged**